File: sample-videochat-webrtc/sample-videochat-webrtc/ViewControllers/CallViewController/CallViewController.m
//
//  CallViewController.m
//  QBRTCChatSemple
//
//  Created by Andrey Ivanov on 11.12.14.
//  Copyright (c) 2014 QuickBlox Team. All rights reserved.
//

#import "CallViewController.h"
#import "LocalVideoView.h"
#import "OpponentCollectionViewCell.h"
#import "OpponentsFlowLayout.h"
#import "QBButton.h"
#import "QBButtonsFactory.h"
#import "QBToolBar.h"
#import "QMSoundManager.h"
#import "Settings.h"
#import "SharingViewController.h"
#import "SVProgressHUD.h"
#import "UsersDataSource.h"
#import "QBCore.h"
#import "StatsView.h"
#import "PlaceholderGenerator.h"
#import "RecordSettings.h"
#import "CallKitManager.h"
#import <AVKit/AVKit.h>

static NSString * const kOpponentCollectionViewCellIdentifier = @"OpponentCollectionViewCellIdentifier";
static NSString * const kSharingViewControllerIdentifier = @"SharingViewController";

static const NSTimeInterval kRefreshTimeInterval = 1.f;

static NSString * const kUnknownUserLabel = @"?";
static NSString * const kQBRTCRecordingTitle = @"[Recording] ";

@interface CallViewController ()

<UICollectionViewDataSource, UICollectionViewDelegateFlowLayout, QBRTCClientDelegate, QBRTCAudioSessionDelegate, LocalVideoViewDelegate>

@property (weak, nonatomic) IBOutlet UICollectionView *opponentsCollectionView;
@property (weak, nonatomic) IBOutlet QBToolBar *toolbar;

@property (strong, nonatomic) NSMutableArray *users;
@property (strong, nonatomic) NSIndexPath *selectedItemIndexPath;

@property (assign, nonatomic) NSTimeInterval timeDuration;
@property (strong, nonatomic) NSTimer *callTimer;
@property (assign, nonatomic) NSTimer *beepTimer;

@property (strong, nonatomic) QBRTCCameraCapture *cameraCapture;
@property (strong, nonatomic) NSMutableDictionary *videoViews;
@property (weak, nonatomic) UIView *zoomedView;

@property (strong, nonatomic) QBButton *dynamicEnable;
@property (strong, nonatomic) QBButton *videoEnabled;
@property (strong, nonatomic) QBButton *audioEnabled;
@property (weak, nonatomic) LocalVideoView *localVideoView;

@property (strong, nonatomic) StatsView *statsView;
@property (assign, nonatomic) BOOL shouldGetStats;

@end

@implementation CallViewController

- (void)dealloc {
    
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    NSLog(@"%@ - %@",  NSStringFromSelector(_cmd), self);
}

- (void)awakeFromNib {
    [super awakeFromNib];
    
    [[QBRTCClient instance] addDelegate:self];
    [[QBRTCAudioSession instance] addDelegate:self];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
    if (!audioSession.isInitialized) {
        [audioSession initializeWithConfigurationBlock:^(QBRTCAudioSessionConfiguration *configuration) {
            // adding blutetooth support
            configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetooth;
            configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowBluetoothA2DP;
            
            // adding airplay support
            configuration.categoryOptions |= AVAudioSessionCategoryOptionAllowAirPlay;
            
            if (_session.conferenceType == QBRTCConferenceTypeVideo) {
                // setting mode to video chat to enable airplay audio and speaker only
                configuration.mode = AVAudioSessionModeVideoChat;
            }
        }];
    }
    
    [self configureGUI];
    
    Settings *settings = [Settings instance];
    
    if (self.session.opponentsIDs.count == 1
        && settings.recordSettings.isEnabled) {
        // recording calls for p2p 1 to 1
        if (self.session.conferenceType == QBRTCConferenceTypeVideo) {
            
            [self.session.recorder setVideoRecordingRotation:settings.recordSettings.videoRotation];
            [self.session.recorder setVideoRecordingWidth:settings.recordSettings.width
                                                   height:settings.recordSettings.height
                                                  bitrate:[settings.recordSettings estimatedBitrate]
                                                      fps:settings.recordSettings.fps];
        }
        NSArray *searchPaths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
        NSString *documentPath = [searchPaths firstObject];
        NSString *filePath = [NSString stringWithFormat:@"%@/file_%f.mp4", documentPath, [NSDate date].timeIntervalSince1970];
        [self.session.recorder startRecordWithFileURL:[NSURL fileURLWithPath:filePath]];
    }
    
    if (self.session.conferenceType == QBRTCConferenceTypeVideo) {
        
#if !(TARGET_IPHONE_SIMULATOR)
        self.cameraCapture = [[QBRTCCameraCapture alloc] initWithVideoFormat:settings.videoFormat
                                                                    position:settings.preferredCameraPostion];
        [self.cameraCapture startSession:nil];
        self.session.localMediaStream.videoTrack.videoCapture = self.cameraCapture;
#endif
    }
    
    self.view.backgroundColor = self.opponentsCollectionView.backgroundColor =
    [UIColor colorWithRed:0.1465 green:0.1465 blue:0.1465 alpha:1.0];
    
    NSMutableArray *users = [NSMutableArray arrayWithCapacity:self.session.opponentsIDs.count + 1];
    [users insertObject:Core.currentUser atIndex:0];
    
    for (NSNumber *uID in self.session.opponentsIDs) {
        
        if (Core.currentUser.ID == uID.integerValue) {
            
            QBUUser *initiator = [self.usersDatasource userWithID:self.session.initiatorID.unsignedIntegerValue];
            
            if (!initiator) {
                
                initiator = [QBUUser user];
                initiator.ID = self.session.initiatorID.integerValue;
            }
            
            [users insertObject:initiator atIndex:0];
            
            continue;
        }
        
        QBUUser *user = [self.usersDatasource userWithID:uID.integerValue];
        if (!user) {
            user = [QBUUser user];
            user.ID = uID.integerValue;
        }
        [users insertObject:user atIndex:0];
    }
    
    self.users = users;
    
    BOOL isInitiator = (Core.currentUser.ID == self.session.initiatorID.unsignedIntegerValue);
    isInitiator ? [self startCall] : [self acceptCall];
    
    self.title = @"Connecting...";
    
    if (CallKitManager.isCallKitAvailable
        && [self.session.initiatorID integerValue] == Core.currentUser.ID) {
        [CallKitManager.instance updateCallWithUUID:_callUUID connectingAtDate:[NSDate date]];
    }
}

- (UIView *)videoViewWithOpponentID:(NSNumber *)opponentID {
    
    if (self.session.conferenceType == QBRTCConferenceTypeAudio) {
        return nil;
    }
    
    if (!self.videoViews) {
        self.videoViews = [NSMutableDictionary dictionary];
    }
    
    id result = self.videoViews[opponentID];
    
    if (Core.currentUser.ID == opponentID.integerValue) {//Local preview
        
        if (!result) {
            
            LocalVideoView *localVideoView = [[LocalVideoView alloc] initWithPreviewlayer:self.cameraCapture.previewLayer];
            self.videoViews[opponentID] = localVideoView;
            localVideoView.delegate = self;
            self.localVideoView = localVideoView;
            
            return localVideoView;
        }
    }
    else {//Opponents
        
        QBRTCRemoteVideoView *remoteVideoView = nil;
        
        QBRTCVideoTrack *remoteVideoTraсk = [self.session remoteVideoTrackWithUserID:opponentID];
        
        if (!result && remoteVideoTraсk) {
            
            remoteVideoView = [[QBRTCRemoteVideoView alloc] initWithFrame:CGRectMake(2, 2, 2, 2)];
            remoteVideoView.videoGravity = AVLayerVideoGravityResizeAspectFill;
            self.videoViews[opponentID] = remoteVideoView;
            result = remoteVideoView;
        }
        
        [remoteVideoView setVideoTrack:remoteVideoTraсk];
        
        return result;
    }
    
    return result;
}

- (void)startCall {
    //Begin play calling sound
    self.beepTimer = [NSTimer scheduledTimerWithTimeInterval:[QBRTCConfig dialingTimeInterval]
                                                      target:self
                                                    selector:@selector(playCallingSound:)
                                                    userInfo:nil
                                                     repeats:YES];
    [self playCallingSound:nil];
    //Start call
    NSDictionary *userInfo = @{@"name" : @"Test",
                               @"url" : @"http.quickblox.com",
                               @"param" : @"\"1,2,3,4\""};
    
    [self.session startCall:userInfo];
}

- (void)acceptCall {
    
    [[QMSoundManager instance] stopAllSounds];
    //Accept call
    NSDictionary *userInfo = @{@"acceptCall" : @"userInfo"};
    [self.session acceptCall:userInfo];
}

- (void)configureGUI {
    
    __weak __typeof(self)weakSelf = self;
    
    if (self.session.conferenceType == QBRTCConferenceTypeVideo) {
        
        self.videoEnabled = [QBButtonsFactory videoEnable];
        [self.toolbar addButton:self.videoEnabled action: ^(UIButton *sender) {
            
            weakSelf.session.localMediaStream.videoTrack.enabled ^=1;
            weakSelf.localVideoView.hidden = !weakSelf.session.localMediaStream.videoTrack.enabled;
        }];
    }
    
    self.audioEnabled = [QBButtonsFactory auidoEnable];
    [self.toolbar addButton:self.audioEnabled action: ^(UIButton *sender) {
        
        weakSelf.session.localMediaStream.audioTrack.enabled ^=1;
        weakSelf.session.recorder.microphoneMuted = !weakSelf.session.localMediaStream.audioTrack.enabled;
    }];
    
    [CallKitManager.instance setOnMicrophoneMuteAction:^{
        weakSelf.audioEnabled.pressed ^= 1;
    }];
    
    if (self.session.conferenceType == QBRTCConferenceTypeAudio) {
        
        self.dynamicEnable = [QBButtonsFactory dynamicEnable];
        [self.toolbar addButton:self.dynamicEnable action:^(UIButton *sender) {
            
            QBRTCAudioDevice device = [QBRTCAudioSession instance].currentAudioDevice;
            
            [QBRTCAudioSession instance].currentAudioDevice =
            device == QBRTCAudioDeviceSpeaker ? QBRTCAudioDeviceReceiver : QBRTCAudioDeviceSpeaker;
        }];
    }
    
    if (self.session.conferenceType == QBRTCConferenceTypeVideo) {
        
        [self.toolbar addButton:[QBButtonsFactory screenShare] action: ^(UIButton *sender) {
            
            SharingViewController *sharingVC =
            [weakSelf.storyboard instantiateViewControllerWithIdentifier:kSharingViewControllerIdentifier];
            sharingVC.session = weakSelf.session;
            
            // put camera capture on pause
            [weakSelf.cameraCapture stopSession:nil];
            
            [weakSelf.navigationController pushViewController:sharingVC animated:YES];
        }];
    }
    
    [self.toolbar addButton:[QBButtonsFactory decline] action: ^(UIButton *sender) {
        
        [weakSelf.callTimer invalidate];
        weakSelf.callTimer = nil;
        
        if (weakSelf.session.recorder.state == QBRTCRecorderStateActive) {
            [SVProgressHUD showWithStatus:NSLocalizedString(@"Saving record", nil)];
            [weakSelf.session.recorder stopRecord:^(NSURL *file) {
                [SVProgressHUD dismiss];
            }];
        }
        [weakSelf.session hangUp:@{@"hangup" : @"hang up"}];
    }];
    
    [self.toolbar updateItems];
    
    // stats reports view
    _statsView = [[StatsView alloc] initWithFrame:self.view.bounds];
    _statsView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    _statsView.hidden = YES;
    [self.view addSubview:_statsView];
    
    // add button to enable stats view
    UIBarButtonItem *statsButton = [[UIBarButtonItem alloc] initWithTitle:@"Stats" style:UIBarButtonItemStylePlain target:self action:@selector(updateStatsView)];
    self.navigationItem.rightBarButtonItem = statsButton;
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    
    [self refreshVideoViews];
    
    if (self.cameraCapture != nil
        && !self.cameraCapture.hasStarted) {
        // ideally you should always stop capture session
        // when you are leaving controller in any way
        // here we should get its running state back
        [self.cameraCapture startSession:nil];
    }
}

- (void)refreshVideoViews {
    
    for (OpponentCollectionViewCell *viewToRefresh  in self.opponentsCollectionView.visibleCells) {
        id v = viewToRefresh.videoView;
        [viewToRefresh setVideoView:nil];
        [viewToRefresh setVideoView:v];
    }
}

#pragma mark - UICollectionViewDataSource

- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    
    return self.users.count;
}

- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    
    OpponentCollectionViewCell *reusableCell = [collectionView
                                                dequeueReusableCellWithReuseIdentifier:kOpponentCollectionViewCellIdentifier
                                                forIndexPath:indexPath];
    QBUUser *user = self.users[indexPath.row];
    NSNumber *userID = @(user.ID);
    
    __weak __typeof(self)weakSelf = self;
    [reusableCell setDidPressMuteButton:^(BOOL isMuted) {
        
        QBRTCAudioTrack *audioTrack = [weakSelf.session remoteAudioTrackWithUserID:userID];
        audioTrack.enabled = !isMuted;
    }];
    
    [reusableCell setVideoView:[self videoViewWithOpponentID:userID]];
    reusableCell.connectionState = [self.session connectionStateForUser:userID];
    
    if (user.ID != [QBSession currentSession].currentUser.ID) {
        
        NSString *title = user.fullName ?: kUnknownUserLabel;
        reusableCell.placeholderImageView.image = [PlaceholderGenerator placeholderWithSize:reusableCell.placeholderImageView.bounds.size title:title];
    }
    
    return reusableCell;
}

#pragma mark - Transition to size

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator {
    [super viewWillTransitionToSize:size withTransitionCoordinator:coordinator];
    
    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext>  _Nonnull context) {
        
        [self refreshVideoViews];
        
    } completion:nil];
}

- (NSIndexPath *)indexPathAtUserID:(NSNumber *)userID {
    
    QBUUser *user = [self.usersDatasource userWithID:userID.unsignedIntegerValue];
    
    if (!user) {
        user = [QBUUser user];
        user.ID = userID.unsignedIntegerValue;
    }
    NSUInteger idx = [self.users indexOfObject:user];
    NSIndexPath *indexPath = [NSIndexPath indexPathForRow:idx inSection:0];
    
    return indexPath;
}

- (void)performUpdateUserID:(NSNumber *)userID block:(void(^)(OpponentCollectionViewCell *cell))block {
    
    NSIndexPath *indexPath = [self indexPathAtUserID:userID];
    OpponentCollectionViewCell *cell = (id)[self.opponentsCollectionView cellForItemAtIndexPath:indexPath];
    block(cell);
}

#pragma mark - Statistic

- (void)updateStatsView {
    self.shouldGetStats ^= 1;
    self.statsView.hidden ^= 1;
}

#pragma mark - QBRTCClientDelegate

- (void)session:(QBRTCSession *)session updatedStatsReport:(QBRTCStatsReport *)report forUserID:(NSNumber *)userID {
    
    NSString *result = [report statsString];
    NSLog(@"%@", result);
    
    // send stats to stats view if needed
    if (_shouldGetStats) {
        [_statsView setStats:result];
        [self.view setNeedsLayout];
    }
}

/**
 * Called in case when you are calling to user, but he hasn't answered
 */
- (void)session:(QBRTCSession *)session userDoesNotRespond:(NSNumber *)userID {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

- (void)session:(QBRTCSession *)session acceptedByUser:(NSNumber *)userID userInfo:(NSDictionary *)userInfo {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

/**
 * Called in case when opponent has rejected you call
 */
- (void)session:(QBRTCSession *)session rejectedByUser:(NSNumber *)userID userInfo:(NSDictionary *)userInfo {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

/**
 *  Called in case when opponent hung up
 */
- (void)session:(QBRTCSession *)session hungUpByUser:(NSNumber *)userID userInfo:(NSDictionary *)userInfo {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

/**
 *  Called in case when receive remote video track from opponent
 */

- (void)session:(QBRTCSession *)session receivedRemoteVideoTrack:(QBRTCVideoTrack *)videoTrack fromUser:(NSNumber *)userID {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            
            QBRTCRemoteVideoView *opponentVideoView = (id)[self videoViewWithOpponentID:userID];
            [cell setVideoView:opponentVideoView];
        }];
    }
}

/**
 *  Called in case when connection is established with opponent
 */
- (void)session:(QBRTCSession *)session connectedToUser:(NSNumber *)userID {
    
    if (session == self.session) {
        
        if (self.beepTimer) {
            
            [self.beepTimer invalidate];
            self.beepTimer = nil;
            [[QMSoundManager instance] stopAllSounds];
        }
        
        if (!self.callTimer) {
            
            if (CallKitManager.isCallKitAvailable
                && [self.session.initiatorID integerValue] == Core.currentUser.ID) {
                [CallKitManager.instance updateCallWithUUID:_callUUID connectedAtDate:[NSDate date]];
            }
            
            self.callTimer = [NSTimer scheduledTimerWithTimeInterval:kRefreshTimeInterval
                                                              target:self
                                                            selector:@selector(refreshCallTime:)
                                                            userInfo:nil
                                                             repeats:YES];
        }
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

/**
 *  Called in case when connection state changed
 */
- (void)session:(QBRTCSession *)session connectionClosedForUser:(NSNumber *)userID {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
            [self.videoViews removeObjectForKey:userID];
            [cell setVideoView:nil];
        }];
    }
}

/**
 *  Called in case when disconnected from opponent
 */
- (void)session:(QBRTCSession *)session disconnectedFromUser:(NSNumber *)userID {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

/**
 *  Called in case when disconnected by timeout
 */
- (void)session:(QBRTCSession *)session disconnectedByTimeoutFromUser:(NSNumber *)userID {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

/**
 *  Called in case when connection failed with user
 */
- (void)session:(QBRTCSession *)session connectionFailedWithUser:(NSNumber *)userID {
    
    if (session == self.session) {
        
        [self performUpdateUserID:userID block:^(OpponentCollectionViewCell *cell) {
            cell.connectionState = [self.session connectionStateForUser:userID];
        }];
    }
}

/**
 *  Called in case when session will close
 */
- (void)sessionDidClose:(QBRTCSession *)session {
    
    if (session == self.session) {
        
        if (CallKitManager.isCallKitAvailable) {
            [CallKitManager.instance endCallWithUUID:_callUUID completion:nil];
        }
        
        if (self.session.recorder.state == QBRTCRecorderStateActive) {
            [SVProgressHUD showWithStatus:NSLocalizedString(@"Saving record", nil)];
            [self.session.recorder stopRecord:^(NSURL *file) {
                [SVProgressHUD dismiss];
            }];
        }
        
        [self.cameraCapture stopSession:nil];
        
        QBRTCAudioSession *audioSession = [QBRTCAudioSession instance];
        if (audioSession.isInitialized
            && ![audioSession audioSessionIsActivatedOutside:[AVAudioSession sharedInstance]]) {
            NSLog(@"Deinitializing QBRTCAudioSession in CallViewController.");
            [audioSession deinitialize];
        }
        
        if (self.beepTimer) {
            
            [self.beepTimer invalidate];
            self.beepTimer = nil;
            [[QMSoundManager instance] stopAllSounds];
        }
        
        [self.callTimer invalidate];
        self.callTimer = nil;
        
        self.toolbar.userInteractionEnabled = NO;
        [UIView animateWithDuration:0.5 animations:^{
            
            self.toolbar.alpha = 0.4;
        }];
        
        self.title = [NSString stringWithFormat:@"End - %@", [self stringWithTimeDuration:self.timeDuration]];
    }
}

#pragma mark - QBRTCAudioSessionDelegate

- (void)audioSession:(QBRTCAudioSession *)audioSession didChangeCurrentAudioDevice:(QBRTCAudioDevice)updatedAudioDevice {
    
    BOOL isSpeaker = updatedAudioDevice == QBRTCAudioDeviceSpeaker;
    if (self.dynamicEnable.pressed != isSpeaker) {
        
        self.dynamicEnable.pressed = isSpeaker;
    }
}

#pragma mark - Timers actions

- (void)playCallingSound:(id)sender {
    
    [QMSoundManager playCallingSound];
}

- (void)refreshCallTime:(NSTimer *)sender {
    
    self.timeDuration += kRefreshTimeInterval;
    NSString *extraTitle = @"";
    if (self.session.recorder.state == QBRTCRecorderStateActive) {
        extraTitle = kQBRTCRecordingTitle;
    }
    self.title = [NSString stringWithFormat:@"%@Call time - %@", extraTitle, [self stringWithTimeDuration:self.timeDuration]];
}

- (NSString *)stringWithTimeDuration:(NSTimeInterval )timeDuration {
    
    NSInteger minutes = timeDuration / 60;
    NSInteger seconds = (NSInteger)timeDuration % 60;
    
    NSString *timeStr = [NSString stringWithFormat:@"%ld:%02ld", (long)minutes, (long)seconds];
    
    return timeStr;
}

- (void)localVideoView:(LocalVideoView *)localVideoView pressedSwitchButton:(UIButton *)sender {
    
    AVCaptureDevicePosition position = self.cameraCapture.position;
    AVCaptureDevicePosition newPosition = position == AVCaptureDevicePositionBack ? AVCaptureDevicePositionFront : AVCaptureDevicePositionBack;
    
    if ([self.cameraCapture hasCameraForPosition:newPosition]) {
        
        CATransition *animation = [CATransition animation];
        animation.duration = .75f;
        animation.timingFunction = [CAMediaTimingFunction functionWithName:kCAMediaTimingFunctionEaseInEaseOut];
        animation.type = @"oglFlip";
        
        if (position == AVCaptureDevicePositionFront) {
            
            animation.subtype = kCATransitionFromRight;
        }
        else if(position == AVCaptureDevicePositionBack) {
            
            animation.subtype = kCATransitionFromLeft;
        }
        
        [localVideoView.superview.layer addAnimation:animation forKey:nil];
        self.cameraCapture.position = newPosition;
    }
}

@end

File: sample-videochat-webrtc-swift/sample-videochat-webrtc-swift/CallViewController.swift
//
//  CallViewController.swift
//  sample-videochat-webrtc-swift
//
//  Created by QuickBlox team
//  Copyright © 2018 QuickBlox. All rights reserved.
//

import UIKit

import Quickblox
import QuickbloxWebRTC
import SVProgressHUD

class CallViewController: UIViewController, QBRTCClientDelegate {
    
    @IBOutlet weak var callBtn: UIButton!
    @IBOutlet weak var logoutBtn: UIBarButtonItem!
    @IBOutlet weak var screenShareBtn: UIButton!
    @IBOutlet weak var endBtn: UIButton!
    @IBOutlet weak var toolbar: UIView!
    @IBOutlet weak var stackView: UIStackView!
    
    open var opponets: [QBUUser]?
    open var currentUser: QBUUser?
    
    var views: [UIView] = []
    var videoCapture: QBRTCCameraCapture!
    var session: QBRTCSession?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        QBRTCClient.initializeRTC()
        QBRTCClient.instance().add(self)
        
        cofigureVideo()
        configureAudio()
        
        self.title = self.currentUser?.login
        self.navigationItem.setHidesBackButton(true, animated:true)
        
        self.screenShareBtn.isHidden = true
        self.endBtn.isHidden = true
        self.toolbar.isHidden = true
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
        
        self.resumeVideoCapture()
    }
    
    //MARK: WebRTC configuration
    
    func cofigureVideo() {
        
        QBRTCConfig.mediaStreamConfiguration().videoCodec = .H264
        
        let videoFormat = QBRTCVideoFormat()
        videoFormat.frameRate = 21
        videoFormat.pixelFormat = .format420f
        videoFormat.width = 640
        videoFormat.height = 480
        
        self.videoCapture = QBRTCCameraCapture(videoFormat: videoFormat, position: .front)
        self.videoCapture.previewLayer.videoGravity = AVLayerVideoGravity.resizeAspectFill
        
        self.videoCapture.startSession {
            
            let localView = LocalVideoView(withPreviewLayer:self.videoCapture.previewLayer)
            self.views.append(localView)
            self.stackView.addArrangedSubview(localView)
        }
    }
    
    func configureAudio() {
        
        QBRTCConfig.mediaStreamConfiguration().audioCodec = .codecOpus
        QBRTCAudioSession.instance().initialize { (configuration: QBRTCAudioSessionConfiguration) -> () in

            var options = configuration.categoryOptions
            if #available(iOS 10.0, *) {
                options = options.union(AVAudioSessionCategoryOptions.allowBluetoothA2DP)
                options = options.union(AVAudioSessionCategoryOptions.allowAirPlay)
            } else {
                options = options.union(AVAudioSessionCategoryOptions.allowBluetooth)
            }
            
            configuration.categoryOptions = options
            configuration.mode = AVAudioSessionModeVideoChat
        }
        QBRTCAudioSession.instance().currentAudioDevice = .speaker
    }
    
    //MARK: Actions

    @IBAction func didPressCamSwitchButton(_ sender: UIButton) {
        toggleButton(button: sender)
        
        self.session?.localMediaStream.videoTrack.isEnabled = !(self.session?.localMediaStream.videoTrack.isEnabled)!
    }
    
    @IBAction func didPressMicSwitchButton(_ sender: UIButton) {
        toggleButton(button: sender)
        
        self.session?.localMediaStream.audioTrack.isEnabled = !(self.session?.localMediaStream.audioTrack.isEnabled)!
    }
    
    @IBAction func didPressDynamicButton(_ sender: UIButton) {
        toggleButton(button: sender)
        
        QBRTCAudioSession.instance().currentAudioDevice =
            QBRTCAudioSession.instance().currentAudioDevice == .speaker ? .receiver : .speaker
    }
    
    @IBAction func didPressCameraRotationButton(_ sender: UIButton) {
        toggleButton(button: sender)
        self.videoCapture.position = self.videoCapture.position == .back ? .front : .back
    }
    
    @IBAction func didPressLogout(_ sender: Any) {
        self.logout()
    }
    
    @IBAction func didPressCall(_ sender: UIButton) {
        
        sender.isHidden = true
        self.endBtn.isHidden = false
        self.toolbar.isHidden = false
        self.logoutBtn.isEnabled = false
        let ids = self.opponets?.map({$0.id})
        self.session = QBRTCClient.instance().createNewSession(withOpponents: ids! as [NSNumber],
                                                              with: .video)
        self.session?.localMediaStream.videoTrack.videoCapture = self.videoCapture
        self.session?.startCall(["info" : "user info"])
    }
    
    @IBAction func didPressEnd(_ sender: UIButton) {
        
        if self.session != nil {
            self.session?.hangUp(nil)
        }
    }
    
    @IBAction func didPressScreenShare(_ sender: UIButton) {
        self.videoCapture.stopSession(nil)
        self.performSegue(withIdentifier: "ScreenShareViewController", sender: self.session)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        let screenShareVC = segue.destination as! ScreenShareViewController
        screenShareVC.session = sender as? QBRTCSession
    }
    
    func didReceiveNewSession(_ session: QBRTCSession, userInfo: [String : String]? = nil) {
        
        if self.session == nil {
            self.session = session
            handleIncomingCall()
        }
    }
    
    func session(_ session: QBRTCBaseSession, connectedToUser userID: NSNumber) {
        
        if (session as! QBRTCSession).id == self.session?.id {
            if session.conferenceType == QBRTCConferenceType.video {
                self.screenShareBtn.isHidden = false
            }
        }
    }
    
    func session(_ session: QBRTCSession, hungUpByUser userID: NSNumber, userInfo: [String : String]? = nil) {
        
        if session.id == self.session?.id {
            self.removeRemoteView(with: userID.uintValue)
            if userID == session.initiatorID {
                self.session?.hangUp(nil)
            }
        }
    }
    
    func session(_ session: QBRTCBaseSession, receivedRemoteVideoTrack videoTrack: QBRTCVideoTrack, fromUser userID: NSNumber) {
        
        if (session as! QBRTCSession).id == self.session?.id {
            
            let remoteView :QBRTCRemoteVideoView = QBRTCRemoteVideoView()
            remoteView.videoGravity = AVLayerVideoGravity.resizeAspect.rawValue
            remoteView.clipsToBounds = true
            remoteView.setVideoTrack(videoTrack)
            remoteView.tag = userID.intValue
            self.views.append(remoteView)
            self.relayout(with: self.views)
        }
    }
    
    func sessionDidClose(_ session: QBRTCSession) {
        
        if session.id == self.session?.id {
            self.callBtn.isHidden = false
            self.endBtn.isHidden = true
            self.toolbar.isHidden = true
            self.logoutBtn.isEnabled = true
            self.screenShareBtn.isHidden = true
            let ids = self.opponets?.map({$0.id})
            for userID in ids! {
                self.removeRemoteView(with: userID)
            }
            self.session = nil
        }
    }
    
    //MARK: Helpers
    
    func toggleButton(button: UIButton) {
        button.isSelected = !button.isSelected
        button.backgroundColor = button.isSelected ?
            UIColor(red: 0.3843, green: 0.3843, blue: 0.3843, alpha: 1.0) :
            UIColor(red: 0.8118, green: 0.8118, blue: 0.8118, alpha: 1.0)
    }
    
    func resumeVideoCapture() {
        // ideally you should always stop capture session
        // when you are leaving controller in any way
        // here we should get its running state back
        if self.videoCapture != nil && !self.videoCapture.hasStarted {
            self.session?.localMediaStream.videoTrack.videoCapture = self.videoCapture
            self.videoCapture.startSession(nil)
        }
    }
    
    func removeRemoteView(with userID: UInt) {
        
        if let i = self.views.index(where: { $0.tag == userID }) {
            self.views.remove(at: i)
            self.relayout(with: self.views)
        }
    }
    
    func relayout(with views: [UIView]) {
        self.stackView.removeAllArrangedSubviews()
        
        for v in views {
            
            if self.stackView.arrangedSubviews.count > 1 {
                let i = views.index(of: v)! % 2
                let s = self.stackView.arrangedSubviews[i] as! UIStackView
                s.addArrangedSubview(v)
            }
            else {
                let hStack = UIStackView()
                hStack.axis = .horizontal
                hStack.distribution = .fillEqually
                hStack.spacing = 5
                hStack.addArrangedSubview(v)
                self.stackView.addArrangedSubview(hStack)
            }
        }
    }
    
    func handleIncomingCall() {
        
        let alert = UIAlertController(title: "Incoming video call", message: "Accept ?", preferredStyle: .actionSheet)
        
        let accept = UIAlertAction(title: "Accept", style: .default) { action in
            self.session?.localMediaStream.videoTrack.videoCapture = self.videoCapture
            self.session?.acceptCall(nil)
            self.callBtn.isHidden = true
            self.endBtn.isHidden = false
            self.toolbar.isHidden = false
            self.logoutBtn.isEnabled = false
        }
        
        let reject = UIAlertAction(title: "Reject", style: .default) { action in
            self.session?.rejectCall(nil)
        }
        
        alert.addAction(accept)
        alert.addAction(reject)
        self.present(alert, animated: true)
    }
    
    func logout() {
        
        SVProgressHUD.show(withStatus: "Logout")
        QBChat.instance.disconnect { _ in
            QBRequest.logOut(successBlock: { _ in
                SVProgressHUD.dismiss()
                self.navigationController?.popViewController(animated: true)
            })
        }
    }
}

extension UIStackView {
    
    func removeAllArrangedSubviews() {
        
        let removedSubviews = arrangedSubviews.reduce([]) { (allSubviews, subview) -> [UIView] in
            self.removeArrangedSubview(subview)
            return allSubviews + [subview]
        }
        // Remove the views from self
        removedSubviews.forEach({ $0.removeFromSuperview() })
    }
}
